text/0162-refactoring-support.md

```diff
@@ -0,0 +1,549 @@
+
+Then the CLI will not be able to establish a 1:1 mapping between the old and new
+names. In this case, it will ask you to confirm the changes:
+
+    Ambiguous Resource Name Changes
+    ┌───┬──────────────────────┐
+    │   │ Resource             │
+    ├───┼──────────────────────┤
+    │ - │ Stack.Queue1A4198146 │
+    │   │ Stack.Queue2B0BA5D32 │
+    ├───┼──────────────────────┤
+    │ + │ Stack.Queue3C7606C37 │
+    │   │ Stack.Queue4D681F510 │
+    └───┴──────────────────────┘
+
+    If you want to take advantage of automatic resource refactoring, avoid 
+    renaming or moving multiple identical resources at the same time.
+
+    If these changes were intentional, and you want to proceed with the
+    resource replacements, please confirm below.
+
+    Do you wish to deploy these changes (y/n)?
+
+To skip this prompt and go straight to deployment, pass the
+`--ignore-ambiguous-refactoring` flag to the CLI, or configure this setting in
+the `cdk.json` file:
+
+```json
+{
+  "app": "...",
+  "ignoreAmbiguousRefactoring": true
+}
+```
+
+If you want to execute only the automatic refactoring, without deploying new
+resources or removing existing ones, you can use the `cdk refactor` command.
+
+### Trigger
+
+This feature will also be available as a [CDK Trigger], which you can enable by
+passing the flag `autoRefactor` to your application:
+
+```typescript
+const app = new App({
+  autoRefactor: true,
+  ...
+});
+```
+
+If this is enabled, the CLI will skip the refactoring step during deployment,
+and a Lambda function will execute it instead, as part of the provisioning
+process by CloudFormation.
+
+---
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `status/api-approved` label was applied to
+the RFC pull request):
+
+```
+
+[ ] Signed-off by API Bar Raiser @rix0rrr
+
+```
+
+## Public FAQ
+
+### What are we launching today?
+
+A new developer experience for CDK users, that allows them to change the
+location of a construct (stack plus logical ID) without causing resource
+replacement. This new experience is available in the CDK CLI `deploy`
+and `refactor` commands, as well as a CDK trigger.
+
+### Why should I use this feature?
+
+If you ever find yourself doing one of the following, you will benefit from
+stack refactoring support:
+
+- Renaming constructs, either intentionally or by mistake.
+- Moving constructs within the same stack. This could be just for better
+  organization, or to create reusable components.
+- Moving constructs between different stacks.
+- Renaming stacks.
+
+### Can the CLI help me resolve ambiguity when refactoring resources?
+
+Not at the moment. One of the constraints we imposed on this feature is that it
+should work in any environment, including CI/CD pipelines, where there is no
+user to answer questions. Although we could easily extend this feature to
+include ambiguity resolution for the interactive case, it wouldn't transfer well
```

Comment:
The core capability is identifying logical ids that changed and generating the refactor for them at run time. Could we add a capability to generate on a developer machine a mapping file that could be used during ci/cd deployment such that the developer can confirm mappings in advance and commit the mapping file to be picked up during CI/CD?

Reply:
We can, but I don't feel comfortable with this. It's just too hard for users to get this right. To create a mapping file, you need to predict what the current state of the environment will be. But when it comes the time to apply this mapping, if the state is not what you predicted, the result is undefined.